File: orbits.jl
# Solve Kepler's equation using Newton-Raphson
function compute_E_NR(M,e; T=Float32) # Union{Vector{Float},Float}
    # initial guess (from Smith (1979))
    E = M + e*sin.(M)./(1 .- sin.(M .+ e) + sin.(M));
    E_old = M*0
    # Thresholds
    orbit_thresh = T(1f-6);
    orbit_max_iter = 50;

    # Newton-Raphson method
    i = 0;
    if (e > 0.5)
        while((i < orbit_max_iter) & (norm(E - E_old) > orbit_thresh))
            i += 1;
            E_old = deepcopy(E);
            E = E_old + (M + e*sin.(E_old) - E_old)./(1 .- e*cos.(E_old));

        end
    else
        while((i < orbit_max_iter) & (norm(E - E_old) > orbit_thresh))
            i += 1;
            E_old = copy(E);
            E = M + e*sin.(E_old);
        end
    end
    return E
end

# Compute coefficients for orbital equations
function compute_coeff(Omega, inclination, omega)
    L1 = cos(Omega)*cos(omega) - sin(Omega)*sin(omega)*cos(inclination);
    M1 = sin(Omega)*cos(omega) + cos(Omega)*sin(omega)*cos(inclination);
    N1 = sin(omega)*sin(inclination);
    L2 = -cos(Omega)*sin(omega) - sin(Omega)*cos(omega)*cos(inclination);
    M2 = -sin(Omega)*sin(omega) + cos(Omega)*cos(omega)*cos(inclination);
    N2 = cos(omega)*sin(inclination);
    return L1, M1, N1, L2, M2, N2
end

# Solve relative binary orbit
function binary_orbit_rel(bparameters,tepoch::Float64)
    Ω = bparameters.Ω*pi/180.; # longitude of ascending node
    i = bparameters.i*pi/180.;
    ω = bparameters.ω*pi/180.; # argument of periapsis
    a = bparameters.a;
    e = bparameters.e;
    # Eccentric anomaly
    E = compute_eccentric_anomaly(bparameters, tepoch);
    # compute orbital coefficients
    L1, M1, N1, L2, M2, N2 = compute_coeff(Ω, i, ω);
    x, y, z = compute_xyz_rel(a, sqrt(1.0 - e^2), e, L1, M1, N1, L2, M2, N2, cos(E), sin(E));
    return 0.0, 0.0, 0.0, x, y, z
end

function binary_orbit_rel_alt(bparameters,tepoch::Float64)
    Ω = bparameters.Ω*pi/180.; # longitude of ascending node
    i = bparameters.i*pi/180.;
    ω = bparameters.ω*pi/180.; # argument of periapsis
    υ = compute_true_anomaly(bparameters, tepoch);
    x = cos(Ω)*cos(ω+υ) - sin(Ω)*sin(ω+υ)*cos(i);
    y = sin(Ω)*cos(ω+υ) - cos(Ω)*sin(ω+υ)*cos(i);
    z = sin(ω+υ)*sin(i);
    return 0.0, 0.0, 0.0, x, y, z
end

# Solve absolute binary orbit
function binary_orbit_abs(bparameters,tepoch::Float64)
    Ω = bparameters.Ω*pi/180.0; # longitude of ascending node
    i = bparameters.i*pi/180.0;
    ω = bparameters.ω*pi/180.0; # argument of periapsis
    q = bparameters.q;
    a = bparameters.a;
    e = bparameters.e;
    υ = compute_true_anomaly(bparameters, tepoch);
    D = a*(1.0 - e^2)./(1.0 .+ e* cos.(υ));  
    # distance of objects from the center of mass
    r1 = D / (1/q+1);
    r2 = D / (1+q);
    L1, M1, N1, L2, M2, N2 = compute_coeff(Ω, i, ω);
    x1, y1, z1 = compute_xyz_abs(L1, M1, N1, L2, M2, N2, υ, r1);
    if ((υ .>= 0.0) & (υ .<= pi))
        x2, y2, z2 = compute_xyz_abs(L1, M1, N1, L2, M2, N2, υ .+pi, r2);
    else
        x2, y2, z2 = compute_xyz_abs(L1, M1, N1, L2, M2, N2, υ .-pi, r2);
    end
    return x1, y1, z1, x2, y2, z2
end

# compute xyz absolute positions
function compute_xyz_abs(L1, M1, N1, L2, M2, N2, nu, r_star)
    astro_north = (L1.*r_star.*cos.(nu) .+ L2.*r_star.*sin.(nu));
    astro_east  = (M1.*r_star.*cos.(nu) .+ M2.*r_star.*sin.(nu));
    astro_z     = (N1.*r_star.*cos.(nu) .+  N2.*r_star.*sin.(nu));
    x = astro_north;
    y = astro_east;
    z = -astro_z;
    return x, y, z
end

# Compute the x,y,z relative positions
function compute_xyz_rel(a, β, e, L1, M1, N1, L2, M2, N2, cos_E, sin_E)
    astro_north = a*(L1*cos_E + β*L2*sin_E - e*L1);
    astro_east  = a*(M1*cos_E + β*M2*sin_E - e*M1);
    astro_z     = a*(N1*cos_E + β*N2*sin_E - e*N1);
    x = astro_north;
    y = astro_east;
    z = -astro_z;
    return x, y, z
end

function compute_separation_alt(bparameters, tepoch) # uses true anomaly
  # dimentionless instantaneous separation of the centers of mass of the two stars
  # Multiply by a to find the real separation
  υ = compute_true_anomaly(bparameters, tepoch);
  e = bparameters.e;
  D = (1 - e^2)./(1 .+ e*cos.(υ)); 
  return D
end

function compute_separation(bparameters, tepoch) 
    # dimentionless instantaneous separation of the centers of mass of the two stars
    # Multiply by a to find the real separation
    E = compute_eccentric_anomaly(bparameters, tepoch);
    e = bparameters.e;
    D = 1 .- e*cos.(E)
    return D
end

# Calculate eccentric anomaly
function compute_eccentric_anomaly(bparameters,tepoch; T=Float32) # Note: this can be called with tepochs = vector
    P = bparameters.P;
    dP = bparameters.dP;
    e = bparameters.e;
    T0 = bparameters.T0;
    # mean angular velocity (mean motion)
    n = T(2pi)/P;
    # Mean anomaly
    if (dP > 1e-12)
        # if orbit is decaying
        M = T(4pi)*(tepoch .- T0) .* (1 ./ (1 .+ sqrt.(1 .+ 2dP*(tepoch .- T0)/P)))/P;
    else
        M = n*(tepoch .- T0);
    end
    # Eccentric anomaly using Newton-Raphson method
    E = mod2pi.(compute_E_NR(M,e));
    return E
end


# Calculate true anomaly
function compute_true_anomaly(bparameters,tepoch)
    E = compute_eccentric_anomaly(bparameters,tepoch)
    e = bparameters.e;
    # calculate true anomaly, following Broucke, R. ; Cefola, P. 
    # A Note on the Relations between True and Eccentric Anomalies in the Two-Body Problem
    # Supposedly avoids numerical issues when the arguments are near ± π, as the two tangents
    # in the classic equation become infinite. 
     β = e/(1+sqrt(1-e^2))
     υ = E .+ 2*atan.((β*sin.(E)) ./ (1 .-  β*cos.(E)))    
    #υ = 2*atan.(sqrt((1+e)/(1-e))*tan.(E/2));    
    return υ
end

# epool = collect(range(0,1.0,length=1000))
# E= collect(range(0,2*pi,length=1000))
# [norm((E .+ 2*atan.((e/(1+sqrt(1-e^2))*sin.(E)) ./ (1 .-  e/(1+sqrt(1-e^2))*cos.(E))))-(2*atan.(sqrt((1+e)/(1-e))*tan.(E/2)))) for e in epool]
# epool = collect(range(0,0.999999,length=1000))
# Epool= collect(range(0,2*pi,length=1000))
# [norm((E .+ 2*atan.((e/(1+sqrt(1-e^2))*sin.(E)) ./ (1 .-  e/(1+sqrt(1-e^2))*cos.(E))))-(2*atan.(sqrt((1+e)/(1-e))*tan.(E/2)))) for e in epool]
# [norm((E .+ 2*atan.((e/(1+sqrt(1-e^2))*sin.(E)) ./ (1 .-  e/(1+sqrt(1-e^2))*cos.(E))))-(2*atan.(sqrt((1+e)/(1-e))*tan.(E/2)))) for e in epool]

# function binary_orbit_vel(bparameters,tepoch::Vector{Float64})
#     Omega = bparameters.long_ascending_node*pi/180.; # longitude of ascending node
#     inclination = bparameters.orbit_incl*pi/180.;
#     omega = bparameters.arg_pariapsis*pi/180.; # argument of periapsis
#     a = bparameters.separation;
#     P = bparameters.P;
#     ecc = bparameters.e;
#     # mean angular velocity (mean motion)
#     n = 2.0 * pi / P;
#     E = compute_eccentric_anomaly(bparameters, tepoch);
#     # velocities
#     L1, M1, N1, L2, M2, N2 = compute_coeff(Omega, inclination, omega);
#     velx = n*(a*sqrt(1.0 - ecc^2)*L2*cos.(E) - a*L1*sin.(E)); # in mas/day
#     vely = n*(a*sqrt(1.0 - ecc^2)*M2*cos.(E) - a*M1*sin.(E));
#     velz = n*(a*sqrt(1.0 - ecc^2)*N2*cos.(E) - a*N1*sin.(E));
#     #distance = bparameters.distance; # in parsecs
#     # conversion to km/s
#     velx *= (pi/1000.0/3600.0/180.0)*3.0857e16/(86400.0);
#     vely *= (pi/1000.0/3600.0/180.0)*3.0857e16/(86400.0);
#     velz *= (pi/1000.0/3600.0/180.0)*3.0857e16/(86400.0);
#     return velx, vely, velz
# end

function compute_masses(bparameters)
    G = 6.67408e-11; # m^3/kg/s^2
    M_sun = 1.98847e30; # kg
    rad_mas = 180/pi*3600*1000
    pc = 3.08567782e16
    q = bparameters.q;
    a = bparameters.a/rad_mas; # radians
    P = bparameters.P*86400.0; # seconds
    d = bparameters.d*pc; # km
    m1 = 4*pi^2*(a*d)^3/((G*P^2*(1+q)))/M_sun; # in solar mass
    m2 = m1*q
    return m1, m2
end

function binary_RV(bparameters, tepoch::Union{Float64, Vector{Float64}}; K1::Float64, K2::Float64, γ::Float64)
    ω = bparameters.ω*pi/180.0; # argument of periapsis
    e = bparameters.e;
    υ = compute_true_anomaly(bparameters, tepoch)
    #calculate radial velocity
    Vrad1 = γ + K1 * (cos.(υ  .+ ω) .+ e * cos(ω))
    Vrad2 = γ + K2 * (cos.(υ .+ (ω .+pi)) .+ e * cos(ω .+ pi))
    return Vrad1, Vrad2
end

function binary_proj_plane(bparameters, tepochs) 
    # Project orbit into the x,y observer plane (= screen plane)
    # Primary is at (0,0)
    Ω = bparameters.Ω*pi/180.0; # longitude of ascending node
    i = bparameters.i*pi/180.0;
    ω = bparameters.ω*pi/180.0; # argument of periapsis
    a = bparameters.a;
    e = bparameters.e;
    υ = compute_true_anomaly(bparameters, tepochs)
    r = a * (1.0 - e^2)./(1.0 .+ e*cos.(υ)); # Separation between the two stars
    O_w = atan.(sin.(υ .+ ω) .* cos(i) , cos.(υ .+ ω))
    θ = O_w .+ Ω;                            # Principal axis angle in cyclindrical coordinates
    ρ = abs.(r .* cos.(υ .+ ω) ./ cos.(O_w)) # Radius in the projected cylindrical coordinates
    x = ρ .* cos.(θ)
    y = ρ .* sin.(θ)
    return x, y, ρ, θ
end

File: rotir_jax/geometry/orbits.py
"""Orbital mechanics for ROTIR - binary star systems.

Implements Keplerian orbit calculations for binary stars:
- Kepler's equation solver (Newton-Raphson)
- Eccentric and true anomaly
- 3D orbital positions (absolute and relative)
- Orbital velocity
- Time-dependent geometry for multi-epoch observations

Key applications:
- Eclipsing binaries
- Spectroscopic binaries
- Visual binaries resolved by interferometry
- Mass transfer systems
- Symbiotic stars

References:
- Murray & Dermott (1999): Solar System Dynamics
- Broucke & Cefola (1972): True/eccentric anomaly relations
"""

import jax
import jax.numpy as jnp
import numpy as np
from typing import Tuple, Optional

# Physical constants
TWO_PI = 2.0 * jnp.pi


def compute_E_NR(
    M: jnp.ndarray,
    e: float,
    max_iter: int = 50,
    thresh: float = 1e-6,
) -> jnp.ndarray:
    """Solve Kepler's equation using Newton-Raphson method.

    Kepler's equation: M = E - e*sin(E)
    where M is mean anomaly, E is eccentric anomaly, e is eccentricity.

    Args:
        M: Mean anomaly (radians)
        e: Orbital eccentricity (0 ≤ e < 1)
        max_iter: Maximum number of iterations (default 50)
        thresh: Convergence threshold (default 1e-6)

    Returns:
        E: Eccentric anomaly (radians)

    Notes:
        - Uses Smith (1979) initial guess for faster convergence
        - Different iteration scheme for high eccentricity (e > 0.5)
        - Vectorized for multiple epochs

    Reference:
        orbits.jl lines 2-26
    """
    # Initial guess from Smith (1979)
    E = M + e * jnp.sin(M) / (1.0 - jnp.sin(M + e) + jnp.sin(M))

    # Newton-Raphson iteration
    if e > 0.5:
        # High eccentricity: full Newton-Raphson
        for _ in range(max_iter):
            E_old = E
            # E_new = E - f(E)/f'(E) where f(E) = E - e*sin(E) - M
            dE = (M + e * jnp.sin(E) - E) / (1.0 - e * jnp.cos(E))
            E = E + dE

            # Check convergence
            if jnp.max(jnp.abs(E - E_old)) < thresh:
                break
    else:
        # Low eccentricity: simplified iteration
        for _ in range(max_iter):
            E_old = E
            E = M + e * jnp.sin(E_old)

            if jnp.max(jnp.abs(E - E_old)) < thresh:
                break

    return E


def compute_eccentric_anomaly(
    P: float,
    e: float,
    T0: float,
    tepoch: jnp.ndarray,
    dP: float = 0.0,
) -> jnp.ndarray:
    """Calculate eccentric anomaly at given epoch(s).

    Args:
        P: Orbital period (days)
        e: Eccentricity
        T0: Time of periastron passage (MJD)
        tepoch: Observation epoch(s) (MJD)
        dP: Period derivative dP/dt for decaying orbits (default 0.0)

    Returns:
        E: Eccentric anomaly (radians), range [0, 2π)

    Notes:
        - Handles orbital decay if dP > 0
        - For circular orbits (e=0): E = M
        - Vectorized for multiple epochs

    Reference:
        orbits.jl lines 134-149
    """
    # Mean motion
    n = TWO_PI / P

    # Mean anomaly
    if dP > 1e-12:
        # Orbital decay case
        dt = tepoch - T0
        M = 4.0 * jnp.pi * dt * (1.0 / (1.0 + jnp.sqrt(1.0 + 2.0 * dP * dt / P))) / P
    else:
        # Standard Keplerian orbit
        M = n * (tepoch - T0)

    # Solve Kepler's equation
    E = compute_E_NR(M, e)

    # Wrap to [0, 2π)
    E = jnp.mod(E, TWO_PI)

    return E


def compute_true_anomaly(
    P: float,
    e: float,
    T0: float,
    tepoch: jnp.ndarray,
    dP: float = 0.0,
) -> jnp.ndarray:
    """Calculate true anomaly from orbital elements.

    Args:
        P: Orbital period (days)
        e: Eccentricity
        T0: Time of periastron passage (MJD)
        tepoch: Observation epoch(s) (MJD)
        dP: Period derivative (default 0.0)

    Returns:
        υ: True anomaly (radians)

    Notes:
        - Uses Broucke & Cefola (1972) formula
        - Avoids numerical issues near υ = ±π
        - More stable than classic tan(υ/2) = sqrt((1+e)/(1-e)) * tan(E/2)

    Reference:
        - Broucke & Cefola (1972): Celestial Mechanics 5, 303
        - orbits.jl lines 152-163
    """
    # Eccentric anomaly
    E = compute_eccentric_anomaly(P, e, T0, tepoch, dP)

    # True anomaly using Broucke & Cefola formula
    # Avoids numerical issues when arguments are near ±π
    beta = e / (1.0 + jnp.sqrt(1.0 - e**2))
    υ = E + 2.0 * jnp.arctan((beta * jnp.sin(E)) / (1.0 - beta * jnp.cos(E)))

    # Alternative classic formula (commented out, less stable):
    # υ = 2.0 * jnp.arctan(jnp.sqrt((1+e)/(1-e)) * jnp.tan(E/2))

    return υ


def compute_orbital_coefficients(
    Omega: float,
    i: float,
    omega: float,
) -> Tuple[float, float, float, float, float, float]:
    """Compute transformation coefficients for orbital equations.

    These coefficients transform from orbital plane to sky plane.

    Args:
        Omega: Longitude of ascending node (radians)
        i: Orbital inclination (radians)
        omega: Argument of periastron (radians)

    Returns:
        L1, M1, N1: Coefficients for radial direction
        L2, M2, N2: Coefficients for tangential direction

    Notes:
        - Convention: x=North, y=East, z=toward observer
        - L,M,N transform orbital (r,θ) to sky (x,y,z)

    Reference:
        orbits.jl lines 28-36
    """
    cos_Omega = jnp.cos(Omega)
    sin_Omega = jnp.sin(Omega)
    cos_i = jnp.cos(i)
    sin_i = jnp.sin(i)
    cos_omega = jnp.cos(omega)
    sin_omega = jnp.sin(omega)

    L1 = cos_Omega * cos_omega - sin_Omega * sin_omega * cos_i
    M1 = sin_Omega * cos_omega + cos_Omega * sin_omega * cos_i
    N1 = sin_omega * sin_i

    L2 = -cos_Omega * sin_omega - sin_Omega * cos_omega * cos_i
    M2 = -sin_Omega * sin_omega + cos_Omega * cos_omega * cos_i
    N2 = cos_omega * sin_i

    return L1, M1, N1, L2, M2, N2


def compute_separation(
    a: float,
    e: float,
    P: float,
    T0: float,
    tepoch: jnp.ndarray,
    dP: float = 0.0,
) -> jnp.ndarray:
    """Compute instantaneous separation between binary components.

    Args:
        a: Semi-major axis (mas)
        e: Eccentricity
        P: Period (days)
        T0: Time of periastron (MJD)
        tepoch: Observation epoch(s) (MJD)
        dP: Period derivative (default 0.0)

    Returns:
        D: Dimensionless separation (actual separation = D * a)

    Notes:
        - D varies with orbital phase
        - At periastron: D = 1 - e
        - At apastron: D = 1 + e
        - For circular orbits: D = 1

    Reference:
        orbits.jl lines 119-124
    """
    E = compute_eccentric_anomaly(P, e, T0, tepoch, dP)
    D = 1.0 - e * jnp.cos(E)
    return D


def compute_separation_from_true_anomaly(
    a: float,
    e: float,
    P: float,
    T0: float,
    tepoch: jnp.ndarray,
) -> jnp.ndarray:
    """Compute separation using true anomaly (alternative method).

    Args:
        a: Semi-major axis (mas)
        e: Eccentricity
        P: Period (days)
        T0: Time of periastron (MJD)
        tepoch: Observation epoch(s) (MJD)

    Returns:
        D: Dimensionless separation

    Notes:
        - Uses r = a(1-e²)/(1+e*cos(υ))
        - Equivalent to compute_separation but via true anomaly

    Reference:
        orbits.jl lines 112-117
    """
    υ = compute_true_anomaly(P, e, T0, tepoch)
    D = (1.0 - e**2) / (1.0 + e * jnp.cos(υ))
    return D


def binary_orbit_relative(
    a: float,
    e: float,
    P: float,
    T0: float,
    Omega: float,
    i: float,
    omega: float,
    tepoch: float,
    dP: float = 0.0,
) -> Tuple[float, float, float, float, float, float]:
    """Compute relative binary orbit (secondary relative to primary).

    Args:
        a: Semi-major axis (mas)
        e: Eccentricity
        P: Period (days)
        T0: Time of periastron (MJD)
        Omega: Longitude of ascending node (degrees)
        i: Inclination (degrees)
        omega: Argument of periastron (degrees)
        tepoch: Observation epoch (MJD)
        dP: Period derivative (default 0.0)

    Returns:
        x1, y1, z1: Position of primary (0, 0, 0 for relative orbit)
        x2, y2, z2: Position of secondary (mas)

    Notes:
        - Primary at origin
        - x: North (RA increasing)
        - y: East (Dec increasing)
        - z: toward observer (negative = away)

    Reference:
        orbits.jl lines 38-47
    """
    # Convert angles to radians
    Omega_rad = jnp.deg2rad(Omega)
    i_rad = jnp.deg2rad(i)
    omega_rad = jnp.deg2rad(omega)

    # Eccentric anomaly
    E = compute_eccentric_anomaly(P, e, T0, jnp.array([tepoch]), dP)[0]

    # Orbital coefficients
    L1, M1, N1, L2, M2, N2 = compute_orbital_coefficients(Omega_rad, i_rad, omega_rad)

    # Compute position
    beta = jnp.sqrt(1.0 - e**2)
    cos_E = jnp.cos(E)
    sin_E = jnp.sin(E)

    x = a * (L1 * cos_E + beta * L2 * sin_E - e * L1)
    y = a * (M1 * cos_E + beta * M2 * sin_E - e * M1)
    z = a * (N1 * cos_E + beta * N2 * sin_E - e * N1)

    # Flip z convention: negative z = toward observer
    z = -z

    return 0.0, 0.0, 0.0, x, y, z


def binary_orbit_absolute(
    a: float,
    e: float,
    P: float,
    T0: float,
    q: float,
    Omega: float,
    i: float,
    omega: float,
    tepoch: float,
    dP: float = 0.0,
) -> Tuple[float, float, float, float, float, float]:
    """Compute absolute binary orbit (both stars around center of mass).

    Args:
        a: Semi-major axis (mas)
        e: Eccentricity
        P: Period (days)
        T0: Time of periastron (MJD)
        q: Mass ratio M2/M1
        Omega: Longitude of ascending node (degrees)
        i: Inclination (degrees)
        omega: Argument of periastron (degrees)
        tepoch: Observation epoch (MJD)
        dP: Period derivative (default 0.0)

    Returns:
        x1, y1, z1: Position of primary (mas)
        x2, y2, z2: Position of secondary (mas)

    Notes:
        - Both stars orbit center of mass
        - Distances: r1 = D/(1/q+1), r2 = D/(1+q)
        - Center of mass at origin
        - q = M2/M1 (mass ratio)

    Reference:
        orbits.jl lines 65-81
    """
    # Convert angles to radians
    Omega_rad = jnp.deg2rad(Omega)
    i_rad = jnp.deg2rad(i)
    omega_rad = jnp.deg2rad(omega)

    # True anomaly
    υ = compute_true_anomaly(P, e, T0, jnp.array([tepoch]), dP)[0]

    # Instantaneous separation
    D = a * (1.0 - e**2) / (1.0 + e * jnp.cos(υ))

    # Distance from center of mass
    r1 = D / (1.0 / q + 1.0)  # Primary
    r2 = D / (1.0 + q)         # Secondary

    # Orbital coefficients
    L1, M1, N1, L2, M2, N2 = compute_orbital_coefficients(Omega_rad, i_rad, omega_rad)

    # Primary position
    x1 = r1 * (L1 * jnp.cos(υ) + L2 * jnp.sin(υ))
    y1 = r1 * (M1 * jnp.cos(υ) + M2 * jnp.sin(υ))
    z1 = r1 * (N1 * jnp.cos(υ) + N2 * jnp.sin(υ))

    # Secondary position (opposite side of COM)
    # Phase shifted by π
    υ2 = υ + jnp.pi if (υ >= 0.0 and υ <= jnp.pi) else υ - jnp.pi

    x2 = r2 * (L1 * jnp.cos(υ2) + L2 * jnp.sin(υ2))
    y2 = r2 * (M1 * jnp.cos(υ2) + M2 * jnp.sin(υ2))
    z2 = r2 * (N1 * jnp.cos(υ2) + N2 * jnp.sin(υ2))

    # Flip z: negative = toward observer
    z1 = -z1
    z2 = -z2

    return x1, y1, z1, x2, y2, z2


def compute_orbital_phase(
    P: float,
    T0: float,
    tepoch: jnp.ndarray,
) -> jnp.ndarray:
    """Compute orbital phase at given epoch(s).

    Args:
        P: Orbital period (days)
        T0: Zero-point epoch (MJD)
        tepoch: Observation epoch(s) (MJD)

    Returns:
        phase: Orbital phase [0, 1], where 0 = T0

    Notes:
        - Phase 0.0 = periastron (if T0 is time of periastron)
        - Phase 0.5 = apastron
        - Useful for phase-folded observations
    """
    phase = jnp.mod((tepoch - T0) / P, 1.0)
    return phase


def compute_radial_velocity(
    K: float,
    e: float,
    P: float,
    T0: float,
    omega: float,
    tepoch: jnp.ndarray,
    gamma: float = 0.0,
) -> jnp.ndarray:
    """Compute radial velocity for spectroscopic binary.

    Args:
        K: Velocity semi-amplitude (km/s)
        e: Eccentricity
        P: Period (days)
        T0: Time of periastron (MJD)
        omega: Argument of periastron (degrees)
        tepoch: Observation epoch(s) (MJD)
        gamma: Systemic velocity (km/s, default 0)

    Returns:
        RV: Radial velocity (km/s)

    Notes:
        - RV = γ + K[cos(ω+υ) + e*cos(ω)]
        - K related to masses and inclination
        - For double-lined: K1*M1 = K2*M2

    Example:
        >>> # Algol-like system
        >>> K = 50.0  # km/s
        >>> tepochs = jnp.linspace(0, 100, 1000)  # 100 days
        >>> RV = compute_radial_velocity(K, 0.1, 2.87, 0.0, 45.0, tepochs)
    """
    # True anomaly
    υ = compute_true_anomaly(P, e, T0, tepoch)

    # Argument of periastron in radians
    omega_rad = jnp.deg2rad(omega)

    # Radial velocity equation
    RV = gamma + K * (jnp.cos(omega_rad + υ) + e * jnp.cos(omega_rad))

    return RV


def compute_masses_from_orbit(
    P: float,
    a: float,
    i: float,
    distance: float,
) -> Tuple[float, float]:
    """Compute total mass from orbital parameters (Kepler's 3rd law).

    Args:
        P: Orbital period (days)
        a: Semi-major axis (mas)
        i: Inclination (degrees)
        distance: Distance to system (pc)

    Returns:
        M_total: Total system mass (solar masses)
        a_au: Semi-major axis (AU)

    Notes:
        - Kepler's 3rd law: M_total = a³/P² (in solar units)
        - Requires distance to convert mas → AU
        - Individual masses need mass ratio q

    Example:
        >>> # Binary at 10 pc with a=100 mas, P=1 year
        >>> M_total, a_au = compute_masses_from_orbit(365.25, 100, 90, 10)
        >>> print(f"Total mass: {M_total:.2f} M_sun")
    """
    # Convert mas to AU
    a_au = a * distance / 1000.0  # mas * pc / 1000 = AU

    # Kepler's 3rd law: P² = a³ / M_total (with P in years, a in AU, M in M_sun)
    P_years = P / 365.25
    M_total = a_au**3 / P_years**2

    return M_total, a_au
